gpu-aware-scheduling/pkg/gpuscheduler/node_resource_cache.go

```diff
@@ -0,0 +1,551 @@
+	annotation, ok := pod.Annotations[cardAnnotationName]
+	if !ok {
+		return
+	}
+
+	item := workQueueItem{
+		ns:         pod.Namespace,
+		name:       pod.Name,
+		annotation: annotation,
+		pod:        pod,
+		action:     podAdded,
+	}
+	c.workQueue.Add(item)
+}
+
+func (c *Cache) updatePodInCache(oldPodObj, newPodObj interface{}) {
+	newPod, ok := newPodObj.(*v1.Pod)
+	if !ok {
+		klog.Warningf("conversion of newObj -> pod failed: %v", newPodObj)
+
+		return
+	}
+
+	// if POD does not have the necessary annotation, can't work on it yet
+	annotation, ok := newPod.Annotations[cardAnnotationName]
+	if !ok {
+		return
+	}
+
+	item := workQueueItem{
+		name:       newPod.Name,
+		ns:         newPod.Namespace,
+		annotation: annotation,
+		pod:        newPod,
+		action:     podUpdated,
+	}
+
+	// Change action to completed if pod is completed
+	if isCompletedPod(newPod) {
+		item.action = podCompleted
+	}
+
+	c.workQueue.Add(item)
+}
+
+func (c *Cache) deletePodFromCache(podObj interface{}) {
+	klog.V(L4).Infof("deletePodFromCache")
+	c.rwmutex.RLock() // reads c.annotatedPods
+	klog.V(L5).Infof("deletePodFromCache locked")
+	defer c.rwmutex.RUnlock()
+
+	var pod *v1.Pod
+	switch t := podObj.(type) {
+	case *v1.Pod:
+		pod = t
+	case cache.DeletedFinalStateUnknown:
+		var ok bool
+		pod, ok = t.Obj.(*v1.Pod)
+
+		if !ok {
+			klog.Warningf("cannot convert to *v1.Pod: %v", t.Obj)
+
+			return
+		}
+	default:
+		klog.Warningf("cannot convert to *v1.Pod: %v", t)
+
+		return
+	}
+
+	key := getKey(pod)
+	_, annotatedPod := c.annotatedPods[key]
+
+	klog.V(L4).Infof("delete pod %s in ns %s annotated:%v", pod.Name, pod.Namespace, annotatedPod)
+
+	if !annotatedPod {
+		return
+	}
+
+	item := workQueueItem{
+		ns:     pod.Namespace,
+		name:   pod.Name,
+		pod:    pod,
+		action: podDeleted,
+	}
+	c.workQueue.Add(item)
+}
+
+// This steals the calling goroutine and blocks doing work.
+func (c *Cache) startWorking(stopChannel <-chan struct{}) {
+	defer c.workQueue.ShutDown()
+	defer runtime.HandleCrash()
+
+	klog.V(L2).Info("Starting worker")
+
+	// block calling goroutine
+	wait.Until(c.workerRun, workerWaitTime, stopChannel)
+
+	klog.V(L2).Info("Worker shutting down")
+}
+
+func (c *Cache) workerRun() {
+	for c.work() {
+	}
+}
+
+func (c *Cache) work() bool {
+	klog.V(L3).Info("worker started")
+
+	itemI, quit := c.workQueue.Get()
+
+	if quit {
+		klog.V(L3).Info("worker quitting")
+
+		return false
+	}
+
+	defer c.workQueue.Done(itemI)
+	defer klog.V(L3).Info("worker ended work")
+
+	item := itemI.(workQueueItem)
+	forget, err := c.handlePod(item)
+
+	if err == nil {
+		if forget {
+			c.workQueue.Forget(itemI)
+		}
+
+		return true
+	}
+
+	klog.Errorf("error handling pod %v ns %v: %v", item.name, item.ns, err)
+	runtime.HandleError(errHandling)
+
+	return true
+}
+
+func getKey(pod *v1.Pod) string {
+	return pod.Namespace + "&" + pod.Name
+}
+
+// this fetches a node by a name.
+func (c *Cache) fetchNode(nodeName string) (*v1.Node, error) {
```

Comment:
Why does this use a node lister instead of making raw API calls?

Reply:
If you mean by raw API calls separate calls to the k8s apiserver, then the answer is reducing apiserver load. The node cache will not need to be updated that often, and there will not need to be calls all the way to the k8s apiserver just for these node fetches.